File: engine/test/parity_test.dart
import 'dart:io';

import 'package:test/test.dart';

/// The parity gate as a test: the Dart engine must reproduce what the Python engine did with
/// the same recorded inputs. See bin/replay.dart and tools/parity/README.md.
void main() {
  final fixtures = Directory('../tools/parity/fixtures')
      .listSync()
      .whereType<Directory>()
      .map((d) => d.path)
      .toList()
    ..sort();

  test('there is at least one recording to replay', () => expect(fixtures, isNotEmpty));

  for (final folder in fixtures) {
    test('replay matches the Python: $folder', () {
      final run = Process.runSync(Platform.resolvedExecutable, ['run', 'bin/replay.dart', folder]);
      expect(run.exitCode, 0, reason: '${run.stdout}\n${run.stderr}');
      expect(run.stdout as String, contains('PARITY OK'));
    }, timeout: const Timeout(Duration(minutes: 5)));
  }
}

File: service/cmd/replay/parity_test.go
package main

import (
	"os"
	"path/filepath"
	"testing"
)

// The parity gate as a test: the Go port of the six strategies must reproduce what the Python
// did with the same recorded inputs. See tools/parity/README.md.
func TestParityWithPython(t *testing.T) {
	fixtures, _ := filepath.Glob("../../../tools/parity/fixtures/*")
	if len(fixtures) == 0 {
		t.Fatal("no recordings to replay")
	}
	for _, folder := range fixtures {
		if st, err := os.Stat(folder); err != nil || !st.IsDir() {
			continue
		}
		t.Run(filepath.Base(folder), func(t *testing.T) {
			if err := run(folder); err != nil {
				t.Fatal(err)
			}
		})
	}
}
